autoload/lsp/omni.vim

```diff
@@ -63,26 +64,65 @@ function! lsp#omni#complete(findstart, base) abort
             while s:completion['status'] is# s:completion_status_pending && !complete_check()
                 sleep 10m
             endwhile
-            let Is_prefix_match = s:create_prefix_matcher(a:base)
-            let s:completion['matches'] = filter(s:completion['matches'], {_, match -> Is_prefix_match(match['word'])})
+
+            " TODO: Allow multiple servers
+            let l:server_name = l:info['server_names'][0]
+            let l:server_info = lsp#get_server_info(l:server_name)
+
+            let l:typed_pattern = has_key(l:server_info, 'config') && has_key(l:server_info['config'], 'typed_pattern') ? l:server_info['config']['typed_pattern'] : '\k*$'
+            let l:current_line = strpart(getline('.'), 0, col('.') - 1)
+            let s:start_pos = match(l:current_line, l:typed_pattern)
+
+            let l:filter = has_key(l:server_info, 'config') && has_key(l:server_info['config'], 'filter') ? l:server_info['config']['filter'] : { 'name': 'none' }
+            let l:last_typed_word = strpart(l:current_line, s:start_pos)
+
+            if l:filter['name'] ==? 'prefix'
+                let s:completion['matches'] = filter(s:completion['matches'], {_, item -> s:prefix_filter(item, l:last_typed_word)})
+            elseif l:filter['name'] ==? 'contains'
+                let s:completion['matches'] = filter(s:completion['matches'], {_, item -> s:contains_filter(item, l:last_typed_word)})
+            endif
+
             let s:completion['status'] = ''
-            return s:completion['matches']
+
+            call timer_start(0, function('s:display_completions'))
+
+            return exists('v:none') ? v:none : []
         endif
     endif
 endfunction
 
-function! s:normalize_word(word) abort
-    if &g:ignorecase
-        return tolower(a:word)
+function! s:get_filter_label(item) abort
+    if !has_key(a:item, 'user_data')
+        return trim(a:item['word'])
+    endif
+
+    let l:user_data = json_decode(a:item['user_data'])
+    return trim(get(l:user_data, s:user_data_filtertext_key, a:item['word']))
+endfunction
+
+function! s:prefix_filter(item, last_typed_word) abort
+    let l:label = s:get_filter_label(a:item)
+    let l:match_pattern = '^' . a:last_typed_word
```

Comment:
please use `stridx(l:label, a:last_typed_word) == 0` or  `stridx(tolower(l:label), tolower(a:last_typed_word)) == 0` instead.

Reply:
@mattn Changed. Is this ready to be merged?